api/handlers/videostream.go

```diff
@@ -35,54 +35,340 @@ LICENSE
 package handlers
 
 import (
+	"context"
+	"strconv"
 	"time"
 
 	"github.com/ausocean/openfish/api/api"
+	"github.com/ausocean/openfish/api/ds_client"
+	"github.com/ausocean/openfish/api/model"
+	"github.com/ausocean/openfish/datastore"
 
 	"github.com/gofiber/fiber/v2"
 )
 
 // VideoStreamResult  describes the JSON format for video streams in API responses.
 // Fields use pointers because they are optional (this is what the format URL param is for).
 type VideoStreamResult struct {
-	ID        *int           `json:"id,omitempty"`
-	Datetime  *time.Time     `json:"datetime,omitempty"`
-	Duration  *time.Duration `json:"duration,omitempty"`
-	StreamUrl *string        `json:"streamUrl,omitempty"`
+	ID            *int       `json:"id,omitempty"`
+	StartTime     *time.Time `json:"startTime,omitempty"`
+	EndTime       *time.Time `json:"endTime,omitempty"`
+	StreamUrl     *string    `json:"stream_url,omitempty"`
+	CaptureSource *int64     `json:"capturesource,omitempty"`
+}
+
+// FromVideoStream creates a VideoStreamResult from a model.VideoStream and key, formatting it according to the requested format.
+func FromVideoStream(videoStream *model.VideoStream, id int, format *api.Format) VideoStreamResult {
+	var result VideoStreamResult
+	if format.Requires("id") {
+		result.ID = &id
+	}
+	if format.Requires("start_time") {
+		result.StartTime = &videoStream.StartTime
+	}
+	if format.Requires("end_time") {
+		result.EndTime = videoStream.EndTime
+	}
+	if format.Requires("stream_url") {
+		result.StreamUrl = &videoStream.StreamUrl
+	}
+	if format.Requires("capturesource") {
+		result.CaptureSource = &videoStream.CaptureSource
+	}
+	return result
 }
 
 // GetVideoStreamsQuery describes the URL query parameters required for the GetVideoStreams endpoint.
 type GetVideoStreamsQuery struct {
-	TimeSpan      *string `query:"timespan"`       // Optional. TODO: choose more appropriate type.
-	CaptureSource *int64  `query:"capture_source"` // Optional.
+	CaptureSource *int64 `query:"capturesource"` // Optional.
+	TimeSpan      *struct {
+		Start time.Time `query:"start"`
+		End   time.Time `query:"end"`
+	} `query:"timespan"` // Optional.
 	api.LimitAndOffset
-	api.Format
+}
+
+// CreateVideoStreamBody describes the JSON format required for the CreateVideoStream endpoint.
+//
+// ID is omitted because it is chosen automatically.
+type CreateVideoStreamBody struct {
+	StartTime     time.Time `json:"startTime"`
```

Comment:
Could the TimeSpan struct be used here ?

Reply:
It could, but I think it should be consistent with UpdateVideoStreamBody, since they are both part of our front facing API. UpdateVideoStreamBody has its own reason to not use *TimeSpan - the start and end time are optional and only one may be supplied in a request.